crates/client/gateway/src/server/handler.rs

```diff
@@ -21,37 +21,89 @@ use crate::error::StarknetError;
+            .or_internal_server_error(format!("Retrieving block {block_id}"))?
+            .ok_or(StarknetError::block_not_found())?;
 
-    if let Ok(block) = MadaraBlock::try_from(block.clone()) {
-        let last_l1_confirmed_block =
-            backend.get_l1_last_confirmed_block().or_internal_server_error("Retrieving last l1 confirmed block")?;
+        if let Ok(block) = MadaraBlock::try_from(block.clone()) {
+            let last_l1_confirmed_block =
+                backend.get_l1_last_confirmed_block().or_internal_server_error("Retrieving last l1 confirmed block")?;
 
-        let status = if Some(block.info.header.block_number) <= last_l1_confirmed_block {
-            BlockStatus::AcceptedOnL1
+            let status = if Some(block.info.header.block_number) <= last_l1_confirmed_block {
+                BlockStatus::AcceptedOnL1
+            } else {
+                BlockStatus::AcceptedOnL2
+            };
+
+            let block_provider = ProviderBlock::new(block, status);
+            Ok(create_json_response(hyper::StatusCode::OK, &block_provider))
         } else {
-            BlockStatus::AcceptedOnL2
-        };
+            let block =
+                MadaraPendingBlock::try_from(block).map_err(|e| GatewayError::InternalServerError(e.to_string()))?;
+            let block_provider = ProviderBlockPending::new(block);
+            Ok(create_json_response(hyper::StatusCode::OK, &block_provider))
+        }
+    }
+}
 
-        let block_provider = ProviderBlock::new(block, status);
-        Ok(create_json_response(hyper::StatusCode::OK, &block_provider))
-    } else {
-        let block =
-            MadaraPendingBlock::try_from(block).map_err(|e| GatewayError::InternalServerError(e.to_string()))?;
-        let block_provider = ProviderBlockPending::new(block);
-        Ok(create_json_response(hyper::StatusCode::OK, &block_provider))
+pub async fn handle_get_signature(
+    req: Request<Body>,
+    backend: Arc<MadaraBackend>,
+) -> Result<Response<Body>, GatewayError> {
+    let params = get_params_from_request(&req);
+    let block_id = block_id_from_params(&params).or_internal_server_error("Retrieving block id")?;
+
+    if matches!(block_id, BlockId::Tag(BlockTag::Pending)) {
+        return Err(GatewayError::StarknetError(StarknetError::no_signature_for_pending_block()));
+    }
+
+    let block_info = backend
+        .get_block_info(&block_id)
+        .or_internal_server_error(format!("Retrieving block info for block {block_id}"))?
+        .ok_or(StarknetError::block_not_found())?;
+
+    match block_info {
+        MadaraMaybePendingBlockInfo::Pending(_) => Err(GatewayError::InternalServerError(format!(
+            "Retrieved pending block info from db for non-pending block {block_id}"
+        ))),
+        MadaraMaybePendingBlockInfo::NotPending(block_info) => {
+            let signature =
+                ProviderBlockSignature { block_hash: block_info.block_hash, signature: block_info.signature };
```

Comment:
I wonder if we shouldn't calculate the signature just in time because the sequencer key can be renewed

Reply:
I think this is fine for now, given that the signature has only been very loosely implemented and is only a static value in the chain config. We will want to add better edge case handling once block signing has been fully implemented into Madara though.

Reply:
the chainConfig is not completely static, you can restart the sequencer with a different version, a different sequencer address, etc. but with the same db

Reply:
https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993

> The key pair will be updated every year (unless the key is compromised, in which case we will replace it immediately)

Reply:
This has been implemented :+1: